cmd/migration-managerd/internal/api/workers.go

```diff
@@ -1048,3 +1060,208 @@ func (d *Daemon) configureMigratedInstances(ctx context.Context, instances migra
 		return nil
 	})
 }
+
+func (d *Daemon) checkSourceConnectivity() {
+	log := slog.With(slog.String("method", "checkSourceConnectivity"))
+
+	// TODO: context should be passed from the daemon to all the workers.
+	ctx := context.TODO()
+
+	sources, err := d.source.GetAll(ctx)
+	if err != nil {
+		log.Warn("Failed to get sources from database", logger.Err(err))
+		return
+	}
+
+	for _, src := range sources {
+		// Skip any sources that already have good connectivity or aren't VMware.
+		if src.GetExternalConnectivityStatus() == api.EXTERNALCONNECTIVITYSTATUS_OK || src.SourceType != api.SOURCETYPE_VMWARE {
+			continue
+		}
+
+		s, err := source.NewInternalVMwareSourceFrom(api.Source{
+			Name:       src.Name,
+			DatabaseID: src.ID,
+			SourceType: src.SourceType,
+			Properties: src.Properties,
+		})
+		if err != nil {
+			log.Warn("Failed to create VMWareSource from source", slog.String("source", src.Name), logger.Err(err))
+			continue
+		}
+
+		// Do a basic connectivity check.
+		status, untrustedCert := doBasicConnectivityCheck(s.Endpoint, s.TrustedServerCertificateFingerprint)
+
+		if untrustedCert != nil && s.ServerCertificate == nil {
+			// We got an untrusted certificate; if one hasn't already been set, add it to this source.
+			s.ServerCertificate = untrustedCert.Raw
+			src.SetServerCertificate(untrustedCert)
+		}
+
+		if status == api.EXTERNALCONNECTIVITYSTATUS_TLS_CONFIRM_FINGERPRINT {
+			// Need to wait for user to confirm if the fingerprint is trusted or not.
+			src.SetExternalConnectivityStatus(status)
+		} else if status != api.EXTERNALCONNECTIVITYSTATUS_OK {
+			// Some other basic connectivity issue occurred.
+			src.SetExternalConnectivityStatus(status)
+		} else {
+			// Basic connectivity is good, now test authentication.
+
+			// Test the connectivity of this source.
+			src.SetExternalConnectivityStatus(mapErrorToStatus(s.Connect(ctx)))
+		}
+
+		// Update the connectivity status in the database.
+		_, err = d.source.UpdateByID(ctx, src)
```

Comment:
Initially I was a bit confused about why we store connectivity status in the database, since it's possible to be out of date if there was a network hiccup at the time of registering the source/target. 

But actually, seems more like a _"has this source/target passed authentication before?"_ state that's being saved, which is fair enough, though _"connectivity"_ to me implies whether the source/target is actively reachable over the network.